pkg/terraform/errors/errors.go

```diff
@@ -16,40 +16,169 @@ limitations under the License.
 
 package errors
 
+import (
+	"fmt"
+	"strings"
+
+	jsoniter "github.com/json-iterator/go"
+	"github.com/pkg/errors"
+)
+
+const (
+	levelError = "error"
+)
+
+type tfError struct {
+	message string
+}
+
 type applyFailed struct {
-	log string
+	*tfError
+}
+
+// TerraformLog represents relevant fields of a Terraform CLI JSON-formatted log line
+type TerraformLog struct {
+	Level      string        `json:"@level"`
+	Message    string        `json:"@message"`
+	Diagnostic LogDiagnostic `json:"diagnostic"`
+}
+
+// LogDiagnostic represents relevant fields of a Terraform CLI JSON-formatted
+// log line diagnostic info
+type LogDiagnostic struct {
+	Severity string `json:"severity"`
+	Summary  string `json:"summary"`
+	Detail   string `json:"detail"`
+	Range    Range  `json:"range"`
+}
+
+// Range represents a line range in a Terraform workspace file
+type Range struct {
+	FileName string `json:"filename"`
+}
+
+func (t *tfError) Error() string {
+	return t.message
 }
 
-func (a *applyFailed) Error() string {
-	return a.log
+func newTFError(message string, logs []byte) (string, *tfError) {
+	tfError := &tfError{
+		message: message,
+	}
+
+	tfLogs, err := parseTerraformLogs(logs)
+	if err != nil {
+		return err.Error(), tfError
+	}
+
+	messages := make([]string, 0, len(tfLogs))
+	for _, l := range tfLogs {
+		// only use error logs
+		if l == nil || l.Level != levelError {
+			continue
+		}
+		m := l.Message
+		if l.Diagnostic.Severity == levelError && l.Diagnostic.Summary != "" {
+			m = fmt.Sprintf("%s: %s", l.Diagnostic.Summary, l.Diagnostic.Detail)
+			if len(l.Diagnostic.Range.FileName) != 0 {
+				m = m + ": File name: " + l.Diagnostic.Range.FileName
+			}
+		}
+		messages = append(messages, m)
+	}
+	tfError.message = fmt.Sprintf("%s: %s", message, strings.Join(messages, "\n"))
+	return "", tfError
+}
+
+func parseTerraformLogs(logs []byte) ([]*TerraformLog, error) {
+	logLines := strings.Split(string(logs), "\n")
+	tfLogs := make([]*TerraformLog, 0, len(logLines))
+	for _, l := range logLines {
+		log := &TerraformLog{}
+		l := strings.TrimSpace(l)
+		if l == "" {
+			continue
+		}
+		if err := jsoniter.ConfigCompatibleWithStandardLibrary.UnmarshalFromString(l, log); err != nil {
+			return nil, err
+		}
+		tfLogs = append(tfLogs, log)
+	}
+	return tfLogs, nil
 }
 
 // NewApplyFailed returns a new apply failure error with given logs.
-func NewApplyFailed(log string) error {
-	return &applyFailed{log: log}
+func NewApplyFailed(logs []byte) error {
+	parseError, tfError := newTFError("apply failed", logs)
+	result := &applyFailed{tfError: tfError}
+	if parseError == "" {
+		return result
+	}
+	return errors.WithMessage(result, parseError)
 }
 
 // IsApplyFailed returns whether error is due to failure of an apply operation.
 func IsApplyFailed(err error) bool {
-	_, ok := err.(*applyFailed)
-	return ok
+	r := &applyFailed{}
```

Comment:
nit: now it's possible to get away with simple type casting.

Reply:
This I also considered. However, in case these errors are themselves wrapped, then the current implementations are more robust. In my opinion, as a principle, we had better always implement similar methods with wrapping in mind.